### PR TITLE
Implement SSO + JWT scenario

### PR DESCRIPTION
**Pull request by Sgitario**
This is the same example as in security JWT, but using Keycloak as token issuer for a webapp application (quarkus.oidc.application-type=web-app).

### CODE REVIEW

File: security/keycloak-jwt/src/test/java/io/quarkus/ts/openshift/security/keycloak/AbstractSecurityKeycloakOpenShiftIT.java [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+import java.net.URL;
+import java.nio.file.Files;
+import java.nio.file.Paths;
+import java.util.List;
+
+public abstract class AbstractSecurityKeycloakOpenShiftIT extends AbstractSecurityKeycloakOpenShiftTest {
+
+    static String keycloakUrl;
+
+    // TODO this is pretty ugly, but I'm tired and can't think of a better way at the moment
```
Comment (Ladicek) [inline security/keycloak-jwt/src/test/java/io/quarkus/ts/openshift/security/keycloak/AbstractSecurityKeycloakOpenShiftIT.java:24]: I'm still ashamed of writing this code, but I can't think of a better way :-) We should perhaps at least encapsulate it into a `static` method somewhere in the `common` module. Not needed in this PR IMHO.

**Comment by Sgitario** [comment]
I've already executed the Jenkins job:
- jvm: https://quarkus-qe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/openshift-test-suite-ci/47/
- native: https://quarkus-qe-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/job/openshift-test-suite-native-ci/20/
Both jobs worked fine.

File: security/keycloak-jwt/src/test/java/io/quarkus/ts/openshift/security/keycloak/containers/KeycloakQuarkusTestResource.java [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+import org.testcontainers.containers.BindMode;
+import org.testcontainers.containers.GenericContainer;
+import org.testcontainers.containers.wait.strategy.Wait;
+
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+
+public class KeycloakQuarkusTestResource implements QuarkusTestResourceLifecycleManager {
```
Comment (Ladicek) [inline security/keycloak-jwt/src/test/java/io/quarkus/ts/openshift/security/keycloak/containers/KeycloakQuarkusTestResource.java:13]: Don't we have this duplicated? Perhaps move to the `common` module.

File: security/keycloak-jwt/src/test/java/io/quarkus/ts/openshift/security/keycloak/AbstractSecurityKeycloakOpenShiftIT.java (hunk 1 above)
Comment (Sgitario) [reply security/keycloak-jwt/src/test/java/io/quarkus/ts/openshift/security/keycloak/AbstractSecurityKeycloakOpenShiftIT.java:24]: There is already an issue to refactor this code: https://github.com/quarkus-qe/quarkus-openshift-test-suite/issues/127

File: security/keycloak-jwt/src/test/java/io/quarkus/ts/openshift/security/keycloak/containers/KeycloakQuarkusTestResource.java (hunk 2 above)
Comment (Sgitario) [reply security/keycloak-jwt/src/test/java/io/quarkus/ts/openshift/security/keycloak/containers/KeycloakQuarkusTestResource.java:13]: For me, It's ok to move this resource into the `common` module. Are you ok with this, @rsvoboda ?

File: security/keycloak-jwt/src/test/java/io/quarkus/ts/openshift/security/keycloak/AbstractSecurityKeycloakOpenShiftIT.java (hunk 1 above)
Comment (Ladicek) [reply security/keycloak-jwt/src/test/java/io/quarkus/ts/openshift/security/keycloak/AbstractSecurityKeycloakOpenShiftIT.java:24]: Ah, right, sorry I forgot.

File: security/keycloak-jwt/src/test/java/io/quarkus/ts/openshift/security/keycloak/containers/KeycloakQuarkusTestResource.java (hunk 2 above)
Comment (Sgitario) [reply security/keycloak-jwt/src/test/java/io/quarkus/ts/openshift/security/keycloak/containers/KeycloakQuarkusTestResource.java]: @Ladicek and @rsvoboda I've moved the Keycloak test resource into the common module. Let me know your thoughts about the current approach.

File: README.md [hunk 3]
```diff
@@ -307,6 +307,85 @@ When a test fails, the test suite will copy the pod logs in order to troubleshoo
...
+<dependency>
+    <groupId>org.testcontainers</groupId>
+    <artifactId>testcontainers</artifactId>
+    <scope>test</scope>
+</dependency>
+```
+
+And also, we'll need to configure the `keycloak-realm.json` in the test classpath of our module. See the next realm configuration as an example:
+
+```json
```
Comment (Sgitario) [inline README.md:339]: Maybe, adding the realm configuration is too much. What do you think?

File: README.md [hunk 4]
```diff
@@ -307,6 +307,85 @@ When a test fails, the test suite will copy the pod logs in order to troubleshoo
...
+      ]
+    }
+  ]
+}
+```
+
+Finally, the Keycloak test resource can be used to auto-configure the Quarkus application:
+
+- Using `KeycloakQuarkusTestResource.WithOidcConfig` to configure the OIDC Auth Url property.
+- Using `KeycloakQuarkusTestResource.WithOidcAndTokenIssuerConfig` to configure the OIDC Auth Url and the Token Issuer properties.
```
Comment (Sgitario) [inline README.md:387]: There are more approaches to do this:
a) Populate all the properties, for all the scenarios, always. 
b) Custom implementation for the KeycloakQuarkusTestResource in each module
c) Different child implementations that populate only a minimal configuration. 

I implemented the c) approach because a) would add too many properties with possible edge cases when running the tests; and b) will still duplicate some code.
What do you think?

File: common/pom.xml [hunk 5]
```diff
@@ -54,6 +54,16 @@
...
         </dependency>
+        <dependency>
+            <groupId>io.quarkus</groupId>
+            <artifactId>quarkus-junit5</artifactId>
+            <scope>provided</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.testcontainers</groupId>
+            <artifactId>testcontainers</artifactId>
+            <scope>provided</scope>
```
Comment (Sgitario) [inline common/pom.xml:65]: I marked this dependency as provided, so the testcontainers dependency do not be added to all the modules. (I mention this in the README.md).

**Comment by Sgitario** [comment]
@rsvoboda can you take a look into this PR? Thanks!

**Comment by rsvoboda** [comment]
My pipeline for today is long, this is already in it 😉

**Comment by rsvoboda** [comment]
I think I won't be able to review this in reasonable time.

I will delegate to @pjgg, hope you don't mind too much;)

@Sgitario you did some changes to the PR after you commented that you run your branch on our jenkins / ocp cluster, should it be rerun or the changes were just minor ones?

**Comment by Sgitario** [comment]
> I think I won't be able to review this in reasonable time.
> 
> I will delegate to @pjgg, hope you don't mind too much;)
> 
> @Sgitario you did some changes to the PR after you commented that you run your branch on our jenkins / ocp cluster, should it be rerun or the changes were just minor ones?

The changes only affected to the non OpenShift tests (a refactor in the testcontainers usage for Keycloak), so the pipeline workflow should suffice.

**Comment by Sgitario** [comment]
@pjgg can you take a look into this PR? Thanks!

File: README.md [hunk 6]
```diff
@@ -307,6 +307,85 @@ When a test fails, the test suite will copy the pod logs in order to troubleshoo
...
+
+```xml
+<dependency>
+    <groupId>org.testcontainers</groupId>
+    <artifactId>testcontainers</artifactId>
+    <scope>test</scope>
+</dependency>
+```
+
+And also, we'll need to configure the `keycloak-realm.json` in the test classpath of our module. See the next realm configuration as an example:
```
Comment (pjgg) [inline README.md:337]:
```suggestion
Also, we'll need to ....
```

File: README.md (hunk 3 above)
Comment (pjgg) [reply README.md:339]: Maybe we should create a security Readme, for the details. And leave this readme as an overview. (link this readme with security readme by a "moreDeatils link")

File: common/src/main/java/io/quarkus/ts/openshift/common/resources/KeycloakQuarkusTestResource.java [hunk 7]
```diff
@@ -32,21 +34,42 @@
...
+        @Override
+        protected Map<String, String> providesQuarkusConfiguration() {
+            return Collections.singletonMap(OIDC_AUTH_URL_PROPERTY, oidcAuthUrl());
+        }
+    }
+
+    public static class WithOidcAndTokenIssuerConfig extends WithOidcConfig {
+        @Override
+        protected Map<String, String> providesQuarkusConfiguration() {
+            Map<String, String> properties = new HashMap<>();
```
Comment (pjgg) [inline common/src/main/java/io/quarkus/ts/openshift/common/resources/KeycloakQuarkusTestResource.java:68]:
```suggestion
            Map<String, String> properties = new HashMap<>(super.providesQuarkusConfiguration());
```

**Review by pjgg — CHANGES_REQUESTED** [review]

File: common/src/main/java/io/quarkus/ts/openshift/common/resources/KeycloakQuarkusTestResource.java (hunk 7 above)
Comment (Sgitario) [reply common/src/main/java/io/quarkus/ts/openshift/common/resources/KeycloakQuarkusTestResource.java:68]: Changed

File: README.md (hunk 6 above)
Comment (Sgitario) [reply README.md:337]: Changed.

**Comment by Sgitario** [comment]
I've applied all the changes from @pjgg except https://github.com/quarkus-qe/quarkus-openshift-test-suite/pull/130#discussion_r548460373. Besides, I like this suggestion, but the security keycloak configuration belongs to the common module, so I prefer the configuration is introduced in the same place as the rest. 
Let me know if it looks good for you now!

**Comment by Sgitario** [comment]
@pjgg, @rsvoboda  what is the status of this PR? I would like to merge it in order to make more progress with the test coverage of Keycloak.

**Comment by rsvoboda** [comment]
I think the only pending items is 
> Maybe we should create a security Readme, for the details. And leave this readme as an overview. (link this readme with security readme by a "moreDeatils link")

**Comment by Sgitario** [comment]
> I think the only pending items is
> 
> > Maybe we should create a security Readme, for the details. And leave this readme as an overview. (link this readme with security readme by a "moreDeatils link")

My concern was about having two places where to configure the same framework, but as you agreed with, I've already done this change.

**Comment by Sgitario** [comment]
I've triggered the jobs again:
- jvm: https://main-jenkins-csb-quarkusqe.cloud.paas.psi.redhat.com/job/openshift-test-suite-ci/108/
- native: https://main-jenkins-csb-quarkusqe.cloud.paas.psi.redhat.com/job/openshift-test-suite-native-ci/14/
Both jobs worked fine.

**Comment by Sgitario** [comment]
run tests

**Comment by Sgitario** [comment]
@pjgg can you take another look into this PR? I've already changed the README.

**Comment by Sgitario** [comment]
run tests

**Comment by Sgitario** [comment]
JVM build failed because of this issue: https://github.com/quarkus-qe/quarkus-openshift-test-suite/pull/138
Native build failed because:
- Keycloak-Authz regression issue: reported by https://github.com/quarkusio/quarkus/issues/14318
- Same issue as in JVM in some modules: https://github.com/quarkus-qe/quarkus-openshift-test-suite/pull/138

**Review by pjgg — APPROVED** [review]

**Comment by Sgitario** [comment]
run tests
